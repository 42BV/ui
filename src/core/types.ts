/* istanbul ignore file */

export type Color =
  | 'primary'
  | 'secondary'
  | 'success'
  | 'info'
  | 'warning'
  | 'danger'
  | 'link'
  | 'muted'
  | 'dark'
  | 'light';

export type ModalSize = 'sm' | 'md' | 'lg' | 'xl';

type Side = 'top' | 'right' | 'bottom' | 'left';
type StartOrEnd = 'start' | 'end';
export type TippyPlacement = Side | `${Side}-${StartOrEnd}`;
