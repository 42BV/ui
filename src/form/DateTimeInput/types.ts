// See https://momentjs.com/docs/#/displaying/format/ for formatting options.
export type DateFormat = string | false; // When false it should not be used
export type TimeFormat = string | false; // When false it should not be used
