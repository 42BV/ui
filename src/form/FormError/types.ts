export type OnChange = (hasErrors: boolean) => void;