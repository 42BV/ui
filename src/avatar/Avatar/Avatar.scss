.avatar {
  .img-placeholder {
    width: 1.875rem;
    height: 1.875rem;
    overflow: hidden;
    border-radius: 50%;
    margin-right: 5px;
    vertical-align: middle;
    display: inline-block;
    background: var(--bs-gray-200);
    line-height: 1;
  }

  img {
    width: 100%;
    height: 100%;
  }

  &.avatar-xs {
    .img-placeholder {
      width: 1.25rem;
      height: 1.25rem;
    }
  }

  &.avatar-sm {
    .img-placeholder {
      width: 3.125rem;
      height: 3.125rem;
    }
  }

  &.avatar-md {
    .img-placeholder {
      width: 4.375rem;
      height: 4.375rem;
    }
  }

  &.avatar-lg {
    .img-placeholder {
      width: 5.625rem;
      height: 5.625rem;
    }
  }
}

.dropdown .avatar {
  position: relative;
  margin-top: -0.125rem;
  margin-right: 0.25rem;
}
