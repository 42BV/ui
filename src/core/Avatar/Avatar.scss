.avatar {
  .img-placeholder {
    width: 1.875rem;
    height: 1.875rem;
    overflow: hidden;
    border-radius: 50%;
    margin-right: 5px;
    vertical-align: middle;
    display: inline-block;
    background: var(--bs-gray-200);
    line-height: 1;
  }

  img {
    width: 100%;
    height: 100%;
  }

  &.avatar-xs {
    .img-placeholder {
      width: 1.25rem;
      height: 1.25rem;
    }
  }

  &.avatar-sm {
    .img-placeholder {
      width: 3.125rem;
      height: 3.125rem;
    }
  }

  &.avatar-md {
    .img-placeholder {
      width: 4.375rem;
      height: 4.375rem;
    }
  }

  &.avatar-lg {
    .img-placeholder {
      width: 5.625rem;
      height: 5.625rem;
    }
  }
}

.avatar-stack {
  > .avatar {
    position: relative;
    display: inline-block;
    margin-right: -1.25rem;

    .img-placeholder {
      margin-right: 0;
    }

    &:hover {
      margin-right: 0;
    }

    @include transition(margin-left 0.25s ease, margin-right 0.25s ease);

    &.avatar-xs,
    &.avatar-sm,
    &.avatar-md,
    &.avatar-lg {
      &:hover {
        margin-right: 0;
      }

      @include transition(margin-left 0.25s ease, margin-right 0.25s ease);
    }

    &.avatar-xs {
      margin-right: -0.625rem;
    }

    &.avatar-sm {
      margin-right: -2.5rem;
    }

    &.avatar-md {
      margin-right: -3.75rem;
    }

    &.avatar-lg {
      margin-right: -80px;
    }
  }
}
