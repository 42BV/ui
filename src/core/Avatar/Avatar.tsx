import React, { ReactNode } from 'react';
import classNames from 'classnames';
import { Tooltip } from '../Tooltip/Tooltip';

type Props = {
  /**
   * Image URL to show as avatar.
   */
  src: string;

  /**
   * Element underneath the image.
   */
  children?: ReactNode;

  /**
   * Text that will be shown upon hovering over the image.
   */
  alt: string;

  /**
   * Optional size.
   *
   * @default md
   */
  size?: AvatarSize;

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;
};

export type AvatarSize = 'xs' | 'sm' | 'md' | 'lg';

/**
 * Avatar is a component which shows a circular image with any element underneath.
 * Use it for instance for showing the profile image of a logged-in user.
 */
export function Avatar({ size, className, alt, src, children }: Props) {
  const sizeClass = size ? `avatar-${size}` : null;
  const classes = classNames('avatar', sizeClass, className);

  return (
    <span className={classes}>
      <Tooltip
        placement="top"
        content={alt}
        distance={tooltipDistanceFromSize(size)}
      >
        <span className="img-placeholder">
          <img alt={alt} src={src} />
        </span>
        {children}
      </Tooltip>
    </span>
  );
}

function tooltipDistanceFromSize(size?: AvatarSize): number {
  switch (size) {
    case 'lg':
      return 42;
    case 'md':
      return 32;
    case 'sm':
      return 22;
  }

  return 7;
}
