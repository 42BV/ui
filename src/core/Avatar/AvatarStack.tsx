import React from 'react';
import classNames from 'classnames';

type Props = {
  /**
   * Element underneath the image.
   */
  children: React.ReactNode;

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;
};

/**
 * AvatarStack is a component which shows a row of overlapped avatars.
 * Use it for instance when you want to show multiple thumbnails of users in an organized way.
 */
export default function AvatarStack({ children, className }: Props) {
  return (
    <div className={classNames('avatar-stack', className)}>{children}</div>
  );
}
