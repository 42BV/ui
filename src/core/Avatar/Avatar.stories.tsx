import React from 'react';
import { storiesOf } from '@storybook/react';

import { Avatar } from './Avatar';
import { AvatarStack } from './AvatarStack';
import { Alert } from 'reactstrap';

storiesOf('core/Avatar', module)
  .addParameters({ component: Avatar })
  .addDecorator((Story) => (
    <>
      <Alert color="warning" className="mb-4">
        <p className="mb-0">To be able to use Avatar, you have to add @tippyjs/react to your dependencies:</p>
        <code>npm install --save @tippyjs/react</code>
        <p className="mb-0 mt-2">You also have to add the stylesheet to your project</p>
        <code>@import &apos;tippy.js/dist/tippy.css&apos;;</code>
      </Alert>
      <Story />
    </>
  ))
  .add('default', () => (
    <div className="text-center pt-5">
      <Avatar alt="lg" size="lg" src="https://www.placecage.com/100/100" />
      <Avatar alt="md" size="md" src="https://www.placecage.com/100/100" />
      <Avatar alt="sm" size="sm" src="https://www.placecage.com/100/100" />
      <Avatar alt="xs" size="xs" src="https://www.placecage.com/100/100" />
      <Avatar alt="muted avatar" src="https://www.placecage.com/100/100">
        <small>John Doe</small>
      </Avatar>
    </div>
  ))
  .add('stacked avatars', () => (
    <div className="text-center">
      <AvatarStack>
        <Avatar
          alt="Avatar number 1"
          size="xs"
          src="https://www.placecage.com/100/100"
        />
        <Avatar
          alt="Avatar number 2"
          size="xs"
          src="https://www.placecage.com/100/100"
        />
        <Avatar
          alt="Avatar number 3"
          size="xs"
          src="https://www.placecage.com/100/100"
        />
        <Avatar
          alt="Avatar number 4"
          size="xs"
          src="https://www.placecage.com/100/100"
        />
      </AvatarStack>
    </div>
  ));
