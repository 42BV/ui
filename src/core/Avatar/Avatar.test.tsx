import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';

import Avatar from './Avatar';
import { BootstrapSize } from '../types';

describe('Component: Avatar', () => {
  let avatar: ShallowWrapper;

  function setup({ size }: { size?: BootstrapSize }) {
    avatar = shallow(
      <Avatar
        className="red"
        size={size}
        src="http://lorempixel.com/200/200/"
        alt="A picture of a tomato"
      >
        <h1>Children</h1>
      </Avatar>
    );
  }

  describe('ui', () => {
    test('default size', () => {
      setup({ size: undefined });

      expect(toJson(avatar)).toMatchSnapshot(
        'Component: Avatar => ui => default size'
      );
    });

    test('non default size', () => {
      setup({ size: 'xs' });

      expect(
        avatar
          .find('span')
          .first()
          .props().className
      ).toContain('avatar-xs');
    });
  });
});
