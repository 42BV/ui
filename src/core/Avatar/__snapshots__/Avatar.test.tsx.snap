// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Avatar ui default size 1`] = `
<div>
  <span
    class="avatar red"
  >
    <span
      style="outline: 0;"
      tabindex="0"
    >
      <span
        class="img-placeholder"
      >
        <img
          alt="A picture of a tomato"
          src="http://lorempixel.com/200/200/"
        />
      </span>
      <h1>
        Children
      </h1>
    </span>
  </span>
</div>
`;

exports[`Component: Avatar ui lg size 1`] = `
<div>
  <span
    class="avatar avatar-lg red"
  >
    <span
      style="outline: 0;"
      tabindex="0"
    >
      <span
        class="img-placeholder"
      >
        <img
          alt="A picture of a tomato"
          src="http://lorempixel.com/200/200/"
        />
      </span>
      <h1>
        Children
      </h1>
    </span>
  </span>
</div>
`;

exports[`Component: Avatar ui md size 1`] = `
<div>
  <span
    class="avatar avatar-md red"
  >
    <span
      style="outline: 0;"
      tabindex="0"
    >
      <span
        class="img-placeholder"
      >
        <img
          alt="A picture of a tomato"
          src="http://lorempixel.com/200/200/"
        />
      </span>
      <h1>
        Children
      </h1>
    </span>
  </span>
</div>
`;

exports[`Component: Avatar ui sm size 1`] = `
<div>
  <span
    class="avatar avatar-sm red"
  >
    <span
      style="outline: 0;"
      tabindex="0"
    >
      <span
        class="img-placeholder"
      >
        <img
          alt="A picture of a tomato"
          src="http://lorempixel.com/200/200/"
        />
      </span>
      <h1>
        Children
      </h1>
    </span>
  </span>
</div>
`;

exports[`Component: Avatar ui xs size 1`] = `
<div>
  <span
    class="avatar avatar-xs red"
  >
    <span
      style="outline: 0;"
      tabindex="0"
    >
      <span
        class="img-placeholder"
      >
        <img
          alt="A picture of a tomato"
          src="http://lorempixel.com/200/200/"
        />
      </span>
      <h1>
        Children
      </h1>
    </span>
  </span>
</div>
`;
